### PR TITLE
chore(tracing): improve the detection of unfinished spans

### PR DESCRIPTION
**Pull request by mabdinur**
Adds a warning log for when a tracer is shutdown with unfinished spans.

## Checklist

- [x] The PR description includes an overview of the change
- [x] The PR description articulates the motivation for the change
- [x] The change includes tests OR the PR description describes a testing strategy
- [x] The PR description notes risks associated with the change, if any
- [x] Newly-added code is easy to change
- [x] The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- [x] The change includes or references documentation updates if necessary
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Newly-added code is easy to change
- [x] Release note makes sense to a user of the library
- [x] If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Comment by mabdinur** [comment]
@brettlangdon  Do we need a unit for this change?

File: ddtrace/_trace/processor/__init__.py [hunk 1]
```diff
@@ -385,6 +385,19 @@ def shutdown(self, timeout):
...
+            for t in self._traces.values()
+            for s in t.spans
+            if not s.finished
+        ]
+        log.warning(
+            "Shutting down tracer with %d unfinished spans. "
+            "Spans that have not been terminated will not be sent to Datadog: %s",
+            len(unfinished_spans),
+            ", ".join(unfinished_spans),
+        )
```
Comment (mabdinur) [inline ddtrace/_trace/processor/__init__.py:400]:
```suggestion
        if unfinished_spans:
            log.warning(
                "Shutting down tracer with %d unfinished spans. "
                "Spans that have not been terminated will not be sent to Datadog: %s",
                len(unfinished_spans),
                ", ".join(unfinished_spans),
            )
```

File: ddtrace/_trace/processor/__init__.py [hunk 2]
```diff
@@ -385,6 +385,20 @@ def shutdown(self, timeout):
         # on_span_finish(...) queues span finish metrics in batches of 100.
         # This ensures all remaining counts are sent before the tracer is shutdown.
         self._queue_span_count_metrics("spans_finished", "integration_name", 1)
+        # Log a warning if the tracer is shutdown before spans are finished
+        unfinished_spans = [
+            f"trace_id={s.trace_id} parent_id={s.parent_id} span_id={s.span_id} name={s.name} resource={s.resource} started={s.start}"  # noqa: E501
```
Comment (mabdinur) [inline ddtrace/_trace/processor/__init__.py:390]:
```suggestion
            f"trace_id={s.trace_id} parent_id={s.parent_id} span_id={s.span_id} name={s.name} resource={s.resource} started={s.start} sampling_priority={s.context.sampling_priority}"  # noqa: E501
```

File: ddtrace/_trace/processor/__init__.py [hunk 3]
```diff
@@ -385,6 +385,20 @@ def shutdown(self, timeout):
...
+        unfinished_spans = [
+            f"trace_id={s.trace_id} parent_id={s.parent_id} span_id={s.span_id} name={s.name} resource={s.resource} started={s.start} sampling_priority={s.context.sampling_priority}"  # noqa: E501
+            for t in self._traces.values()
+            for s in t.spans
+            if not s.finished
+        ]
+        if unfinished_spans:
+            log.warning(
+                "Shutting down tracer with %d unfinished spans. "
+                "Spans that have not been terminated will not be sent to Datadog: %s",
```
Comment (erikayasuda) [inline ddtrace/_trace/processor/__init__.py:398]: nit: Maybe?
```suggestion
                "Shutting down tracer with %d unfinished spans. "
                "Unfinished spans will not be sent to Datadog: %s",
```

**Review by erikayasuda — APPROVED** [review]

File: tests/tracer/test_tracer.py [hunk 4]
```diff
@@ -1110,6 +1110,35 @@ def test_enable():
         assert not t2.enabled
 
 
+@pytest.mark.subprocess(
+    err=b"Shutting down tracer with 2 unfinished spans. "
+    b"Spans that have not been terminated will not be sent to Datadog: "
```
Comment (mabdinur) [inline tests/tracer/test_tracer.py:1115]:
```suggestion
    b"Unfinished spans will not be sent to Datadog: "
```

**Pull request merged**